File: src/money.ts
/** Currency fraction digits for ISO 4217 codes used in the app. */
const FRACTION_DIGITS: Record<string, number> = {
  KZT: 2,
  USD: 2,
  EUR: 2,
};

export function currencyFractionDigits(currency: string): number {
  return FRACTION_DIGITS[currency.toUpperCase()] ?? 2;
}

export function scaleForCurrency(currency: string): number {
  const digits = currencyFractionDigits(currency);
  let scale = 1;
  for (let i = 0; i < digits; i += 1) {
    scale *= 10;
  }
  return scale;
}

/**
 * Parse a user-entered amount into signed minor units.
 * Uses integer arithmetic only — never floating money math.
 */
export function parseMoneyInput(input: string, currency: string): number {
  let trimmed = input.trim().replace(/\s/g, "").replace(",", ".");
  if (!trimmed || trimmed === "-" || trimmed === "+" || trimmed === ".") {
    throw new Error("Введите сумму");
  }

  // Allow "12." while typing / before blur.
  if (trimmed.endsWith(".")) {
    trimmed = trimmed.slice(0, -1);
  }
  if (!trimmed || trimmed === "-" || trimmed === "+") {
    throw new Error("Введите сумму");
  }

  const sign = trimmed.startsWith("-") ? -1 : 1;
  const unsigned = trimmed.replace(/^[+-]/, "");
  if (!/^\d+(\.\d+)?$/.test(unsigned)) {
    throw new Error("Некорректная сумма");
  }

  const [wholePart, fractionPart = ""] = unsigned.split(".");
  const digits = currencyFractionDigits(currency);
  if (fractionPart.length > digits) {
    throw new Error(`Максимум ${digits} знака после запятой`);
  }

  const whole = Number.parseInt(wholePart || "0", 10);
  if (!Number.isSafeInteger(whole)) {
    throw new Error("Слишком большая сумма");
  }

  const padded = fractionPart.padEnd(digits, "0");
  const fraction = padded ? Number.parseInt(padded, 10) : 0;
  const scale = scaleForCurrency(currency);
  const minor = whole * scale + fraction;
  if (!Number.isSafeInteger(minor)) {
    throw new Error("Слишком большая сумма");
  }

  return sign * minor;
}

/** Empty or blank limit fields count as zero instead of throwing. */
export function parseMoneyInputOrZero(input: string, currency: string): number {
  const trimmed = input.trim();
  if (!trimmed || trimmed === "." || trimmed === "-" || trimmed === "+") {
    return 0;
  }
  return parseMoneyInput(trimmed, currency);
}

export function formatMoney(amountMinor: number, currency: string, locale = "ru-RU"): string {
  const digits = currencyFractionDigits(currency);
  const scale = scaleForCurrency(currency);
  const absolute = Math.abs(amountMinor);
  const whole = Math.trunc(absolute / scale);
  const fraction = absolute % scale;
  const sign = amountMinor < 0 ? "-" : "";
  const major =
    digits === 0
      ? String(whole)
      : `${whole}.${String(fraction).padStart(digits, "0")}`;

  const asNumber = Number(`${sign}${major}`);
  return new Intl.NumberFormat(locale, {
    style: "currency",
    currency: currency.toUpperCase(),
    minimumFractionDigits: digits,
    maximumFractionDigits: digits,
  }).format(asNumber);
}

export function sumMinor(values: Iterable<number>): number {
  let total = 0;
  for (const value of values) {
    if (!Number.isSafeInteger(value)) {
      throw new Error("Некорректная денежная сумма");
    }
    total += value;
    if (!Number.isSafeInteger(total)) {
      throw new Error("Переполнение при сложении сумм");
    }
  }
  return total;
}

File: crates/finance-telegramd/src/money.rs
fn currency_fraction_digits(currency: &str) -> u32 {
    match currency.to_ascii_uppercase().as_str() {
        "JPY" => 0,
        _ => 2,
    }
}

fn scale_for_currency(currency: &str) -> i64 {
    let digits = currency_fraction_digits(currency);
    let mut scale = 1_i64;
    for _ in 0..digits {
        scale = scale.saturating_mul(10);
    }
    scale
}

/// Parse a user-entered amount into signed minor units (integer only).
pub fn parse_money_input(input: &str, currency: &str) -> Result<i64, String> {
    let mut trimmed = input.trim().replace(' ', "").replace(',', ".");
    if trimmed.is_empty() || trimmed == "-" || trimmed == "+" || trimmed == "." {
        return Err("Введите сумму".into());
    }
    if trimmed.ends_with('.') {
        trimmed.pop();
    }
    if trimmed.is_empty() || trimmed == "-" || trimmed == "+" {
        return Err("Введите сумму".into());
    }

    let sign = if trimmed.starts_with('-') { -1 } else { 1 };
    let unsigned = trimmed.trim_start_matches(['+', '-']);
    let parts: Vec<&str> = unsigned.split('.').collect();
    if parts.is_empty() || parts.len() > 2 || parts[0].is_empty() {
        return Err("Некорректная сумма".into());
    }
    if !parts[0].chars().all(|c| c.is_ascii_digit()) {
        return Err("Некорректная сумма".into());
    }
    if let Some(frac) = parts.get(1) {
        if !frac.chars().all(|c| c.is_ascii_digit()) {
            return Err("Некорректная сумма".into());
        }
    }

    let digits = currency_fraction_digits(currency) as usize;
    let fraction_part = parts.get(1).copied().unwrap_or("");
    if fraction_part.len() > digits {
        return Err(format!("Максимум {digits} знака после запятой"));
    }

    let whole: i64 = parts[0]
        .parse()
        .map_err(|_| "Слишком большая сумма".to_string())?;
    let padded = format!("{:0<width$}", fraction_part, width = digits);
    let fraction: i64 = if digits == 0 {
        0
    } else {
        padded
            .parse()
            .map_err(|_| "Слишком большая сумма".to_string())?
    };
    let scale = scale_for_currency(currency);
    let minor = whole
        .checked_mul(scale)
        .and_then(|v| v.checked_add(fraction))
        .ok_or_else(|| "Слишком большая сумма".to_string())?;
    Ok(sign * minor)
}

pub fn format_money(amount_minor: i64, currency: &str) -> String {
    let digits = currency_fraction_digits(currency) as usize;
    let scale = scale_for_currency(currency).max(1);
    let sign = if amount_minor < 0 { "−" } else { "" };
    let abs = amount_minor.abs();
    let whole = abs / scale;
    if digits == 0 {
        return format!("{sign}{whole} {currency}");
    }
    let frac = abs % scale;
    format!("{sign}{whole}.{frac:0width$} {currency}", width = digits)
}
